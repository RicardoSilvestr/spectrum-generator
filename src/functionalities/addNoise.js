import random from 'random';
import seedrandom from 'seedrandom';
/**
  * Add noise to the spectrum
  * @param {object} [data] - Your spectrum data in the format {x:[x1, x2, ...], y:[y1, y2, ...]}
  * @param {number} [percent = 0] - Noise's amplitude in percents of the spectrum max value
  * @param {object} [options={}]
  * @param {boolean} [options.seed=false] - Deterministic sequence of random number
  * @param {string} [options.distribution='uniform'] -Type of random: 'uniform' (true random), 'normal' (gaussian distribution),
  * @return {data} data
  */
export default function addNoise(data, percent = 0, options = {}) {
  const {
    distribution = 'uniform',
    seed = false
  } = options;

  if (seed) random.use(seedrandom(0));

  let generateRandomNumber;
  switch (distribution) {
    case 'uniform':
      generateRandomNumber = random.uniform(-0.5, 0.5);
      break;
    case 'normal':
      generateRandomNumber = random.normal();
      break;
    default:
      throw new Error(`Unknown distribution ${options.distribution}`);
  }

  if (!percent) return data;
  var ys = data.y;
  var factor = percent * findMax(ys) / 100;
  for (let i = 0; i < ys.length; i++) {
    ys[i] += generateRandomNumber() * factor;
  }
  return data;
}

function findMax(array) {
  let max = Number.MIN_VALUE;
  for (let item of array) {
    if (item > max) max = item;
  }
  return max;
}
