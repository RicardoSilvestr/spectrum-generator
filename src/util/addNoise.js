import { randomUniform, randomNormal } from 'd3-random';
import XSAdd from 'ml-xsadd';

export default function addNoise(data, percent = 0, options = {}) {
  const { distribution = 'uniform', seed } = options;

  let generateRandomNumber;
  switch (distribution) {
    case 'uniform': {
      generateRandomNumber = getRandom(randomUniform, seed, -0.5, 0.5);
      break;
    }
    case 'normal': {
      generateRandomNumber = getRandom(randomNormal, seed);
      break;
    }
    default:
      throw new Error(`Unknown distribution ${options.distribution}`);
  }

  if (!percent) return data;
  let ys = data.y;
  let factor = (percent * findMax(ys)) / 100;
  for (let i = 0; i < ys.length; i++) {
    ys[i] += generateRandomNumber() * factor;
  }
  return data;
}

function getRandom(func, seed, ...args) {
  return typeof seed === 'number'
    ? func.source(new XSAdd(seed).random)(...args)
    : func(...args);
}

function findMax(array) {
  let max = Number.MIN_VALUE;
  for (let item of array) {
    if (item > max) max = item;
  }
  return max;
}
