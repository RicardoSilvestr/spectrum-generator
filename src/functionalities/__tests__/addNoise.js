import addNoise from '../addNoise';

describe('Test addNoise', () => {
  var corrected = addNoise({ x: [1, 2, 3], y: [2, 3, 4] }, 10, { seed: true });
  expect(corrected.x).toHaveLength(3);
  expect(corrected).toMatchSnapshot();
});

