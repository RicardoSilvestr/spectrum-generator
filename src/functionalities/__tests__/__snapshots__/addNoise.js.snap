// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "x": Array [
    1,
    2,
    3,
  ],
  "y": Array [
    1.8152341483105883,
    2.8344985399814133,
    4.036841599485438,
  ],
}
`;
